python/cudf/cudf/_lib/pylibcudf/sorting.pyx

```diff
@@ -0,0 +1,351 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+
+from libcpp.memory cimport unique_ptr
+from libcpp.utility cimport move
+from libcpp.vector cimport vector
+
+from cudf._lib.cpp cimport sorting as cpp_sorting
+from cudf._lib.cpp.aggregation cimport rank_method
+from cudf._lib.cpp.column.column cimport column
+from cudf._lib.cpp.table.table cimport table
+from cudf._lib.cpp.types cimport null_order, null_policy, order
+
+from .column cimport Column
+from .table cimport Table
+
+
+cpdef Column sorted_order(Table source_table, list column_order, list null_precedence):
+    """Computes the row indices required to sort the table.
+
+    Parameters
+    ----------
+    source_table : Table
+        The table to sort.
+    column_order : List[ColumnOrder]
+        Whether each column should be sorted in ascending or descending order.
+    null_precedence : List[NullOrder]
+        Whether nulls should come before or after non-nulls.
+
+    Returns
+    -------
+    Column
+        The row indices required to sort the table.
+    """
+    cdef unique_ptr[column] c_result
+    cdef vector[order] c_orders = column_order
+    cdef vector[null_order] c_null_precedence = null_precedence
+    with nogil:
+        c_result = move(
+            cpp_sorting.sorted_order(
+                source_table.view(),
+                c_orders,
+                c_null_precedence,
+            )
+        )
+    return Column.from_libcudf(move(c_result))
+
+
+cpdef Column stable_sorted_order(
```

Comment:
I agree that's a nicer approach in the long term. I wasn't focusing so much on that in the first pass, but no harm in making these improvements now rather than later.

Reply:
Ah I forgot, there was a reason I didn't do this. cdef functions don't support this (and therefore neither do cpdef functions). I think we tried this in one of the previous pylibcudf PRs.

For now the plan is to expose the entire API via cpdef functions to allow both typed and untyped usage of the API. We can always consider revisiting that decision in the future if we see a reason to.